### DISCUSSION TITLE
Amend a typing decision?

### DISCUSSION THREAD

**Discussion by gwhitney** [DanielXMoore/Civet#1006] [General]
By what I would consider a series of historical accidents, in TypeScript,  if `A` is of type `T[]` and `n` is of type `number` then `A[n]` is of type number but `A.at(n)` is of type number|undefined. It's as though TypeScript thinks that "ooh, negative numbers are _hard_, you're much more likely to mess up your indexing with at." So as a result I end up with ugly type assertions or lots of `or DummyT`s all over my code in places where it's convenient to work with negative indices, such as when the elements are naturally indexed by $\mathbb Z_n$.

Any chance Civet could/would translate `A.at(n)` to `(A.at(n) as T)`? (Any reason that would be a really bad idea? I couldn't see one, which is why I posted.)

**Comment by edemaine** [comment]
I wonder if Civet is the right level to fix this. Fundamentally, it's a TS lib issue. So I wonder if it makes more sense to override the es2022 lib that introduced `at` using an NPM patch, and possibly [TS's lib override support](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#supporting-lib-from-node_modules).

At a practical level, it's hard for Civet to fix this, because it would need to know the type of `A` and realize that it's deriving from this particular lib declaration... (not just `Array`s, but things that extend it, etc.)

**Reply by gwhitney** [reply]
> At a practical level, it's hard for Civet to fix this, because it would need to know the type of `A` and realize that it's deriving from this particular lib declaration... (not just `Array`s, but things that extend it, etc.)

Fair enough. The path of least resistance for me is just to define an operator `arr at i` that's properly typed. Since I do this a lot, I'd like to make it a single character, which is my original motivation for #1034. (Unfortunately `@` is taken ;-) but Unicode is big enough to find something comfortable.)

**Reply by gwhitney** [reply]
P.S. I am happy to close this discussion or have you do it but didn't think I should unilaterally even though it seems I can...
